File: contest-management-cs/contest-management-gui-cs/repository/ICompetitionRepository.cs
using System.Collections.Generic;
using contest_management_gui_cs.model;

namespace contest_management_gui_cs.repository
{
    public interface ICompetitionRepository : ICrudRepository<long, Competition>
    {
        IDictionary<Competition, int> CountParticipantsForEachCompetition();
        
        IEnumerable<string> FindTypesByString(string competitionType);

        IEnumerable<string> FindAgeCategories(string competitionType);

        Competition FindByProps(string competitionType, string ageCategory);
    }
}

File: contest-management-client-server-java/ContestMgmtPersistence/src/main/java/contestmgmt/persistence/repository/ICompetitionRepository.java
package contestmgmt.persistence.repository;

import contestmgmt.model.Competition;
import contestmgmt.model.Tuple;

import java.util.Map;

public interface ICompetitionRepository extends ICrudRepository<Long, Competition> {
    public Map<Long, Tuple<Competition, Integer>> countParticipantsForEachCompetition(String competitionType,
                                                                                      String ageCategory);

    Iterable<Competition> findCompetitionsByString(String competitionType, String ageCategory);

    Iterable<String> findAgeCategoriesFromCompetitionType(String competitionType);

    Competition findByProps(String competitionType, String ageCategory);
}
